src/rules/fileNameCasingRule.ts

```diff
@@ -28,6 +28,74 @@ enum Casing {
     SnakeCase = "snake-case",
 }
 
+type RegexConfig = Record<string, Casing>;
+
+type SimpleConfig = Casing;
+
+type Config = SimpleConfig | RegexConfig;
+
+interface ValidatorSuccessfulResult {
+    valid: true;
+}
+
+interface ValidatorFailureResult {
+    valid: false;
+    failedCasing: Casing;
+}
+
+type ValidatorResult = ValidatorSuccessfulResult | ValidatorFailureResult;
+
+type Validator<T> = (casing: T) => (sourceFile: ts.SourceFile) => ValidatorResult;
```

Comment:
> `Validator<T> = (casing: T) => ...`

If you really want to be type safe here, you could consider `Validator<T extends Config>`.

Then again, TypeScript is pretty good at inferencing local variable types. You could get rid of `Validator` and explicitly type validator methods. That seems slightly preferable to get rid of this extra line.

Reply:
> Then again, TypeScript is pretty good at inferencing local variable types. You could get rid of Validator and explicitly type validator methods. That seems slightly preferable to get rid of this extra line.

I am not following this, I improved the types but what do you want me to remove?!

Reply:
You can remove the `type Validator<T> = > ...` line, and instead of this:

```typescript
const RegexValidator: Validator<RegexConfig> = casingConfig => (sourceFile): ValidatorResult => {
```

...you can do this:

```typescript
const regexValidator = (casingConfig: Casing, sourceFile: ts.SourceFile):  ValidatorResult => {
```

_(along with whatever other names / return types are changed)_

It's a little less code overall.

Reply:
@JoshuaKGoldberg oh got it.

Well, I prefer to have a type that forces the validators to implement that type.

Right now, it will require us to add the same types to `simpleConfigValidator` and` regexValidator` so multiple places.

Also, I believe is more boilerplate code overall actually, since you need to add the types for all the variables and return instead of just doing `Validator<T>`

Reply:
> prefer to have a type

Interesting. Why? Normally, manually declared types such as these are meant for when you have to pass variables around, and need to restrict external (other file, function, etc.) constructs from allowing the wrong type. You only ever use the `Validator` functions in one place, and if you were to type them incorrectly, TypeScript would give a compiler error.

> more boilerplate code

Not really. Instead of this:

```typescript
type Validator<T extends Config> = (sourceFile: ts.SourceFile, casing: T) => ValidationResult;
const validateWithRegexConfig: Validator<RegexConfig> = (sourceFile, casingConfig) => {
const validateWithSimpleConfig: Validator<Casing> = (sourceFile, casingConfig) => {
```

You'd get this:

```typescript
const validateWithRegexConfig: (sourceFile: ts.SourceFile, casingConfig: RegexConfig) => {
const validateWithSimpleConfig: (sourceFile: ts.SourceFile, casingConfig: Casing) => {
```

...which is 2 lines instead of 3, and about the same length on the `const` lines.

This isn't blocking for the PR going in. More of a preference and a curiosity.

Reply:
@JoshuaKGoldberg well, you are missing `ValidationResult` from the second example, unless you are assuming that TypeScript infers the return types for you, which fair enough but if you do that then anything that I return from those functions are valid, I am not forced to return `Casing | undefined` which is what you must return.

When my read the type alone, it has more meaning, which is the same reason why I cast one type `Casing` into another type `SimpleConfig` and why I created simple types like `ValidationResult`, 

I shaped the application design into meaniful types that helps others to read and reason about the code without having to understand the internals.

This is very beneficial when you are looking one type at the same, as you would do in your IDE/Editors. 

If you use the definition suggested, fair enough, you still have the type checking and system but the types itself are not that meaningful.

When you read `(sourceFile: ts.SourceFile, casing: T) => {};`.

- OK, a function that takes `X, and Y` and no clue about what returns.

You don't think about `Validators` since nothing tells you what that is , **except if you know the source code for sure.** or have the best naming ever 😄 

Which for people that are not familiar with the source code will have a harder time dealing with it, they will need to map too many things in their head at once for be able to understand.

The more code you have, the harder the task become.

But when you read `Validator<RegexConfig>` at least the type system is telling you what is going on for that function, the intentation of it and tracing the types you could understand many things without even looking at the source code of specific functions

Or you could create new validators as long as they follow the same shape (logically is another history).

Without that `Validator` type, you are copy-pasting code, but you do not have a defined boundery, you are hoping that the copy-pasting from the other functions are right 😅, but you have no answers of what is the corelation between them.

Why do they always expect those two params 🤷‍♂️ 

That is my thinking process when it comes to types, I used them for documentation, and a mechanism to help others to understand the code.

If you add `tsdoc` with documentation this become even more valuable.

When you document about `Validators` your documentation will be inlined with the type (the same for `SimpleConfig` bind)

Anyway, habit, as the person that will be using this, I prefer to have more meaniful types than just having type checking.